content/post/2023-11-16-neat-patterns-for-github-merge-queues.md

```diff
@@ -0,0 +1,181 @@
+
+## Running a "second round of tests" on the merge queue only
+
+As mentioned above, having a second round of tests that run only when developers _intend to merge_ their change really helps reduce the turnaround time for a change (and reduce costs when running CI!). Unfortunately, the way GHMQs work, it's not wholly obvious how to do that unless you spent literal days reading their docs (I was working on the GHMQ stuff as a "side project" at work and it took me about 2 weeks to come up with this solution).
+
+The easiest (and only) way I know of to achieve a second round of tests is defining a set of [`jobs:`](https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow) that "gather" up other jobs' outcomes and use some logic to decide if they are relevant in the current "round". Only these "gatherer" jobs get to be required statuses on the branch protection rule. Their dependents can be in turn be conditionally run, depending on the "round".
+
+### Defining jobs that "gather" up other jobs' statuses
+
+GHA jobs can define a list of other jobs they ["need"](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) before they can start. This is the key for our solution. Here's a job that (naively) runs after two other jobs named "tests" and "lints" have run, and if they have passed, passes in turn (naively):
+
+```yaml
+# Beware! This is a naive non-solution!
+jobs:
+  # ...
+  can_merge:
+    needs: [tests, lints]
+    runs-on: "ubuntu-latest"
+    steps:
+    - name: yay
+      run: |
+        echo success
+```
+
+This feels too easy. If you didn't notice the comment, you might have glanced at the scroll bar and noticed that there's a lot of article left to go. Yes. Of course there's a catch.
+
+GitHub Actions that have a skipped action on their "needs" list are skipped in turn. This isn't so bad if "tests" passes but "lints" gets skipped. But if "tests" fails, your job will be skipped, which counts as a successful status! You can add `if: always()` to the job - but then the above job will also run (and pass!) even if any of its dependencies have failed. All that would be a massive violation of the "not rocket science" principle of having a code repo that always passes tests. Oops!
+
+So, what can we do? We're programmers, we can always add more logic. Here's an action that always runs, but when it runs, checks the status of its "needed" jobs & only succeeds if the required criteria hold:
+
+```yaml
+jobs:
+  # ...
+  can_merge:
+    needs: [tests, lints]
+    if: always()
+    permissions:
+      actions: read
+    runs-on: ubuntu-latest
+    steps:
+    - env:
+        NEEDS_JSON: "${{toJSON(needs)}}"
+      name: Transform outcomes
+      run: |
+        echo "ALL_SUCCESS=$(echo "$NEEDS_JSON" | jq '. | to_entries | map([.value.result == "success", .value.result == "skipped"] | any) | all')" >>$GITHUB_ENV
```

Comment:
Nice article @antifuchs! This snippet reminded me of an action I made related to a similar problem, so I figured I'd share: https://github.com/marketplace/actions/alls-green#why.